### PR TITLE
Close #152 Represent Bools as Enums

### LINKED ISSUE #152: Representation of Bool Types
@PrometheusPi reported a problem reading particle `bool` attributes with `h5py` in https://github.com/ComputationalRadiationPhysics/picongpu/issues/682.

The problem is of course, that HDF5 has [no native boolean type](http://www.hdfgroup.org/hdf5-quest.html#bool) why we implemented a ~~1~~ 8 byte (?) bitfield as a work-around.

`h5py` instead uses [HDF5 enums](http://docs.h5py.org/en/latest/faq.html) as a work around that will then be directly mapped to a 1-byte numpy bool while reading, see also the [HDF5 Enum Docs](http://www.hdfgroup.org/HDF5/doc/H5.user/Datatypes.html#Datatypes_Enum) (section 8).

The idea now is to also switch to that representation to increase compatibility (that will increase the internal file format version).

### PR DESCRIPTION
**Pull request by ax3l**
Proposal to change the internal representation of `bool` values to be compatible with `h5py` as documented in #152.

- increases the major file format version since that won't be backwards compatible, e.g., for restarts from files

### To Do

- [x] add `h5py` compatibility tests to `.travis.yml` / `./run_tests`
- [x] double-check `sizeof (bool)` which is `1 Byte` vs. real size of `8 Bytes` after conversation - correct in all cases (add a read/write test?)
- [x] actually this is not an API change so `1.2.5` milestone would be ok, too. But it looks a bit odd to make such a change in a `revision` - so go for the `1.3` target version?
- [x] add enum/bool read support (e.g., for restarts)

### CODE REVIEW

**Comment by ax3l** [comment]
@PrometheusPi can you check out this libSplash branch and compile PIConGPU against it to run and test your example again?

**Comment by PrometheusPi** [comment]
@ax3l  Yes, I will check this out later today. Would be great if it works.

**Comment by PrometheusPi** [comment]
I complied PIConGPU with libSplash from this pull. When writing the checkpoint of time step `0`, the following exception was thrown:
```
[1,0]<stderr>:terminate called after throwing an instance of 'splash::DCException'
[1,0]<stderr>:  what():  Exception for DCDataSet [radiationFlag] create: Failed to create dataset
[1,0]<stderr>:[kepler003:08315] *** Process received signal ***
[1,0]<stderr>:[kepler003:08315] Signal: Aborted (6)
[1,0]<stderr>:[kepler003:08315] Signal code:  (-6)
[1,0]<stderr>:[kepler003:08315] [ 0] /lib/x86_64-linux-gnu/libpthread.so.0(+0x10340) [0x7f3f12921340]
[1,0]<stderr>:[kepler003:08315] [ 1] /lib/x86_64-linux-gnu/libc.so.6(gsignal+0x39) [0x7f3f10bc2cc9]
[1,0]<stderr>:[kepler003:08315] [ 2] /lib/x86_64-linux-gnu/libc.so.6(abort+0x148) [0x7f3f10bc60d8]
[1,0]<stderr>:[kepler003:08315] [ 3] /opt/pkg/compiler/gnu/gcc/4.6.2/lib64/libstdc++.so.6(_ZN9__gnu_cxx27__verbose_terminate_handlerEv+0x11d) [0x7f3f11227c6d]
[1,0]<stderr>:[kepler003:08315] [ 4] /opt/pkg/compiler/gnu/gcc/4.6.2/lib64/libstdc++.so.6(+0xbedd6) [0x7f3f11225dd6]
[1,0]<stderr>:[kepler003:08315] [ 5] /opt/pkg/compiler/gnu/gcc/4.6.2/lib64/libstdc++.so.6(+0xbee03) [0x7f3f11225e03]
[1,0]<stderr>:[kepler003:08315] [ 6] /opt/pkg/compiler/gnu/gcc/4.6.2/lib64/libstdc++.so.6(+0xbeefe) [0x7f3f11225efe]
[1,0]<stderr>:[kepler003:08315] [ 7] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN6splash9DCDataSet6createERKNS_14CollectionTypeEiNS_10DimensionsEjbb+0x92c) [0x76866c]
[1,0]<stderr>:[kepler003:08315] [ 8] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN6splash21ParallelDataCollector12writeDataSetEiNS_10DimensionsES1_RKNS_14CollectionTypeEjNS_9SelectionEPKcPKv+0x146) [0x751926]
[1,0]<stderr>:[kepler003:08315] [ 9] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN6splash21ParallelDataCollector5writeEiNS_10DimensionsES1_RKNS_14CollectionTypeEjNS_9SelectionEPKcPKv+0x28c) [0x755e3c]
[1,0]<stderr>:[kepler003:08315] [10] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN6splash23ParallelDomainCollector11writeDomainEiNS_10DimensionsES1_RKNS_14CollectionTypeEjNS_9SelectionEPKcNS_6DomainENS_16IDomainCollector12DomDataClassEPKv+0x121) [0x75d181]
[1,0]<stderr>:[kepler003:08315] [11] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN6splash23ParallelDomainCollector11writeDomainEiRKNS_14CollectionTypeEjNS_9SelectionEPKcNS_6DomainES7_NS_16IDomainCollector12DomDataClassEPKv+0x235) [0x75c905]
[1,0]<stderr>:[kepler003:08315] [12] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN8picongpu4hdf517ParticleAttributeINS_24placeholder_definition1813radiationFlagEEclIN5PMacc5FrameINS_23OperatorCreateVectorBoxENS6_19ParticleDescriptionIN5boost3mpl6stringILi101ELi0ELi0ELi0ELi0ELi0ELi0ELi0EEENS6_4math2CT6VectorIN4mpl_10integral_cIiLi8EEESJ_NSH_2naEEENSB_6v_itemINS_24placeholder_definition1313globalCellIdxINS_24placeholder_definition2017globalCellIdx_picENS6_24placeholder_definition1113pmacc_isAliasEEENSM_IS3_NSM_INS_24placeholder_definition1613momentumPrev1ENSM_INS_24placeholder_definition179weightingENSM_INS_24placeholder_definition158momentumENSM_INS_24placeholder_definition128positionINS_24placeholder_definition1412position_picESS_EENSB_7vector0ISK_EELi0EEELi0EEELi0EEELi0EEELi0EEELi0EEENSB_6vectorINS_24placeholder_definition2214particlePusherINS_9particles6pusher5BorisESS_EENS_24placeholder_definition215shapeINS1G_6shapes3TSCESS_EENS_24placeholder_definitio??????Y ? Q????E???E?@???
[1,0]<stderr>:??? ? Q????E???E?{ { ?Y???????????(A???,??3@=????[kepler003:08315] [13] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN8picongpu4hdf512WriteSpeciesINS_9ParticlesIN5PMacc19ParticleDescriptionIN5boost3mpl6stringILi101ELi0ELi0ELi0ELi0ELi0ELi0ELi0EEENS3_4math2CT6VectorIN4mpl_10integral_cIiLi8EEESE_NSC_2naEEENS6_6v_itemINS_24placeholder_definition1813radiationFlagENSH_INS_24placeholder_definition1613momentumPrev1ENSH_INS_24placeholder_definition179weightingENSH_INS_24placeholder_definition158momentumENSH_INS_24placeholder_definition128positionINS_24placeholder_definition1412position_picENS3_24placeholder_definition1113pmacc_isAliasEEENS6_7
[1,0]<stderr>:vector0ISF_EELi0EEELi0EEELi0EEELi0EEELi0EEENS6_6vectorINS_24placeholder_definition2214particlePusherINS_9particles6pusher5BorisESV_EENS_24placeholder_definition215shapeINS17_6shapes3TSCESV_EENS_24placeholder_definition2313interpolationINS_28FieldToParticleInterpolationIS1E_NS_30AssignedTrilinearInterpolationEEESV_EENS_24placeholder_definition247currentINS_13curre??????Y ? Q????E???E?@???
[1,0]<stderr>:??? ? Q????E???E?{ { ?Y?????[kepler003:08315] [14] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN8picongpu4hdf510HDF5Writer9writeHDF5EPv+0x6f1) [0x711831]
[1,0]<stderr>:[kepler003:08315] [15] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN8picongpu4hdf510HDF5Writer20notificationReceivedEjb+0x36c) [0x71206c]
[1,0]<stderr>:[kepler003:08315] [16] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN5PMacc16SimulationHelperILj2EE11dumpOneStepEj+0x121) [0x6ab471]
[1,0]<stderr>:[kepler003:08315] [17] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN5PMacc16SimulationHelperILj2EE15startSimulationEv+0x130) [0x6a6d40]
[1,0]<stderr>:[kepler003:08315] [18] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(_ZN8picongpu17SimulationStarterINS_21InitialiserControllerENS_16PluginControllerENS_12MySimulationEE5startEv+0x72) [0x6f6a42]
[1,0]<stderr>:[kepler003:08315] [19] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu(main+0x65) [0x652945]
[1,0]<stderr>:[kepler003:08315] [20] /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf5) [0x7f3f10badec5]
[1,0]<stderr>:[kepler003:08315] [21] /data/hplsim/production/HHG_frontSurfaceInstability/011_Kluge_A_radiation_libSplashFix/picongpu/bin/picongpu() [0x652def]
[1,0]<stderr>:[kepler003:08315] *** End of error message ***
```

**Comment by ax3l** [comment]
fixed, let us add the `h5py` read example now on `h5/testWriteRead_0_0_0.h5` data: `deep/folders/data_bool`

**Comment by ax3l** [comment]
@PrometheusPi tests added and it should work now. can you please give it a try with your PIConGPU example again?

**Comment by PrometheusPi** [comment]
@ax3l I am already on it.

**Comment by ax3l** [comment]
@f-schmitt-zih ready for review! :)

File: src/include/splash/basetypes/ColTypeBool.hpp [hunk 1]
```diff
@@ -35,8 +35,11 @@ class ColTypeBool : public CollectionType
 
     ColTypeBool()
     {
-        const hsize_t dim[] = {sizeof (bool)};
-        this->type = H5Tarray_create(H5T_NATIVE_B8, 1, dim);
+        this->type = H5Tenum_create(H5T_NATIVE_INT8);
```
Comment (ax3l) [inline src/include/splash/basetypes/ColTypeBool.hpp]: 8-bit native int [link](http://www.hdfgroup.org/HDF5/doc/RM/PredefDTypes.html)
- [ ] add comment above?

**Comment by PrometheusPi** [comment]
:+1: great job @ax3l - it works !!!

**Comment by ax3l** [comment]
awesome, thanks!

File: tests/SimpleDataTest.cpp [hunk 2]
```diff
@@ -147,6 +169,15 @@ bool SimpleDataTest::subtestWriteRead(Dimensions gridSize, Dimensions borderSize
             resultsCorrect = false;
             break;
         }
+        if (boolRead[i] != boolWrite[i])
+        {
+#if defined TESTS_DEBUG
+            std::cout << i << ": " << boolRead[i] << " != exptected " << boolWrite[i] << std::endl;
```
Comment (f-schmitt-zih) [inline tests/SimpleDataTest.cpp]: typo, "expected"

File: tests/SimpleDataTest.cpp [hunk 3]
```diff
@@ -147,6 +169,15 @@ bool SimpleDataTest::subtestWriteRead(Dimensions gridSize, Dimensions borderSize
...
+#endif
+            resultsCorrect = false;
+            break;
+        }
+    }
 
     delete[] dataRead;
```
Comment (f-schmitt-zih) [inline tests/SimpleDataTest.cpp]: delete boolRead here

File: tests/SimpleDataTest.cpp [hunk 4]
```diff
@@ -157,8 +188,12 @@ bool SimpleDataTest::subtestWriteRead(Dimensions gridSize, Dimensions borderSize
     // locations in a cleared array (-1)
 
     dataRead = new uint64_t[bufferSize];
+    boolRead = new bool[bufferSize];
     for (uint64_t i = 0; i < bufferSize; i++)
+    {
         dataRead[i] = UINT64_MAX;
+        boolRead[i] = false;
```
Comment (f-schmitt-zih) [inline tests/SimpleDataTest.cpp]: you create boolRead a second time but than don't actually read the data?

File: tests/include/SimpleDataTest.h [hunk 5]
```diff
@@ -34,8 +34,8 @@ class SimpleDataTest : public CPPUNIT_NS::TestFixture
 {
     CPPUNIT_TEST_SUITE(SimpleDataTest);
 
-    CPPUNIT_TEST(testWriteRead);
     CPPUNIT_TEST(testNullWrite);
+    CPPUNIT_TEST(testWriteRead);
```
Comment (f-schmitt-zih) [inline tests/include/SimpleDataTest.h]: Any reason for that change? :) I liked my order

File: tests/run_tests [hunk 6]
```diff
@@ -26,6 +26,7 @@ function testMPI()
 }
 
 testSerial ./SimpleDataTest.cpp.out "Testing simple data read/write..."
+testSerial ./readBool.py "Testing h5py compatible read..."
```
Comment (f-schmitt-zih) [inline tests/run_tests]: Would it make sense to check if python is available and run the readBool.py only in that case?

File: src/include/splash/basetypes/ColTypeBool.hpp [hunk 7]
```diff
@@ -35,8 +35,15 @@ class ColTypeBool : public CollectionType
...
     {
-        const hsize_t dim[] = {sizeof (bool)};
-        this->type = H5Tarray_create(H5T_NATIVE_B8, 1, dim);
+        // 8 bit (very) short int, see
+        //   http://www.hdfgroup.org/HDF5/doc/RM/PredefDTypes.html
+        // this is a h5py compatible implementation for bool, see:
+        //   http://docs.h5py.org/en/latest/faq.html
+        this->type = H5Tenum_create(H5T_NATIVE_INT8);
+        const char *names[2] = {"true", "false"};
+        int64_t val[2] = {1, 0};
```
Comment (f-schmitt-zih) [inline src/include/splash/basetypes/ColTypeBool.hpp]: You can make this `const`, too.

**Comment by f-schmitt-zih** [comment]
Good job!

File: tests/SimpleDataTest.cpp (hunk 2 above)
Comment (ax3l) [inline tests/SimpleDataTest.cpp]: Ha, copy paste typo! Thanks! :)

Am 10. Februar 2015 06:29:56 MEZ, schrieb f-schmitt-zih <notifications@github.com>:

File: tests/include/SimpleDataTest.h (hunk 5 above)
Comment (ax3l) [inline tests/include/SimpleDataTest.h]: Yes, I want to read the data later on via h5py :)
the two tests use the same file (overwrite).

Am 10. Februar 2015 06:32:53 MEZ, schrieb f-schmitt-zih <notifications@github.com>:

File: tests/SimpleDataTest.cpp (hunk 4 above)
Comment (ax3l) [inline tests/SimpleDataTest.cpp]: true, I will remove that. that is dead code from the dev stage.

File: tests/run_tests (hunk 6 above)
Comment (ax3l) [inline tests/run_tests]: yes, the same is true for the `CppUnit` dependency as we got as a feedback from our sys admin.

There are actually some issues for the upcoming [1.3](https://github.com/ComputationalRadiationPhysics/libSplash/issues?q=is%3Aopen+is%3Aissue+milestone%3A%22Version+1.3%22) release that will take care of that.

File: src/include/splash/basetypes/ColTypeBool.hpp (hunk 7 above)
Comment (ax3l) [inline src/include/splash/basetypes/ColTypeBool.hpp]: added, thx!

File: tests/SimpleDataTest.cpp (hunk 3 above)
Comment (ax3l) [inline tests/SimpleDataTest.cpp]: done, thx!

**Comment by ax3l** [comment]
@f-schmitt-zih thank you!
Just I added you review comments a minute ago :)

**Comment by ax3l** [comment]
I also added testing with a simple, *serial* h5py read now.

note: if someone wants to test parallel hdf5 reads, there are several issues in `h5py` (https://github.com/h5py/h5py/issues/542) and `pip` (https://github.com/pypa/pip/issues/18) that make automated installs a bit tricky.

**Comment by ax3l** [comment]
@f-schmitt-zih ready for final review/merge if you like. Run-Time tests are all clean :)

**Comment by bussmann** [comment]
Great job!

**Pull request merged**